File: 78. Subsets.py
class Solution:
	def subsets(self, nums: List[int]) -> List[List[int]]:
		res = []

		def backtrack(arr, index) -> None:
			if index > len(nums):
				return

			# make a shallow copy
			res.append(arr[:])

			for j in range(index, len(nums)):
				arr.append(nums[j])
				backtrack(arr, j + 1)
				arr.pop()

		backtrack([], 0)
		return res

# time: O(n * 2^n) n is the length of arr
# spacE: O(n)

File: 78. Subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(nums, 0, res, new ArrayList<Integer>());
        return res;
    }

    // helper function to generate all possible subsets
    public void dfs(int[] nums, int startIndex, List<List<Integer>> res, List<Integer> subset) {
        // append the subset being constructed so far into the result
        res.add(new ArrayList<>(subset)); // create a new copy
        for (int i = startIndex; i < nums.length; i++) {
            subset.add(nums[i]);
            dfs(nums, i + 1, res, subset);
            subset.remove(subset.size() - 1);
        }

    }

    // time: O(n * 2^n), n is length of arr
    // space: O(n), how much the recursive calls take
}